lib/Transforms/GenStateStruct.cpp

```diff
@@ -0,0 +1,1021 @@
+      OpBuilder builder, Value operand, InstanceGraph &instanceGraph,
+      SmallVector<std::pair<mlir::Operation *, LLVM::LLVMPointerType>>
+          &externTypeCollect,
+      hw::InstanceOp inst, LLVM::GEPOp gep, LLVM::ConstantOp zeroC);
+
+  /// The function to convert hw::InOutType to LLVM compatible type.
+  /// Attention: we support InOutType / nested ArrayType /nested
+  /// UnpackedArrayType to convert to LLVM compatible Type. Other nested
+  /// InOutType will hit the assertions in program and stop execution of
+  /// this pass.
+  Type convertInOutType(hw::InOutType type);
+  LLVM::LLVMStructType getExternModulePortTy(MLIRContext *ctx, bool noOpaqueTy,
+                                             HWModuleExternOp op);
+
+public:
+  GenStateStruct(bool emitNoneInfo) : emitNoneInfo(emitNoneInfo) {}
+  /// The function to add prefix to the given name and return the new name.
+  std::string addPrefixToName(StringRef name);
+
+  /// This function is used to generate the struct type based on the given
+  /// infomation of current module. The struct type is packed defaultly.
+  /// If the module is an extern module which only have a declaration, generate
+  /// opaque struct type for it.
+  LLVM::LLVMStructType createStructType(MLIRContext *ctx, HWModuleOp op,
+                                        InstanceGraph &instGraph,
+                                        SmallVector<PortInfoCollect> &pInfo,
+                                        SmallVector<RegInfoCollect> &rInfo);
+
+  /// The function to create the global struct that is used to store all state
+  /// values and ports infomation.
+  LLVM::GlobalOp createGlobalStruct(Location loc, OpBuilder &builder,
+                                    ModuleOp module, StringRef name,
+                                    LLVM::LLVMStructType type);
+
+  /// The function to initialize the global struct with default value 0, and
+  /// return initialized struct object.
+  void initializeGlobalStruct(OpBuilder &builder, HWModuleOp op,
+                              LLVM::GlobalOp globalOp,
+                              LLVM::LLVMStructType type,
+                              InstanceGraph &instGraph,
+                              SmallVector<PortInfoCollect> &pInfo,
+                              SmallVector<RegInfoCollect> &rInfo);
+
+  /// The function is used to collect all ports info above the given module.
+  /// The collected info: port name, port type, port direction and whether it is
+  /// trigger signal.
+  void collectPortInfo(HWModuleOp op, SmallVector<PortInfoCollect> &pInfo);
+
+  /// The function is used to collect all regs info in the given module.
+  /// The collected info: reg name, reg type and whether it is output reg.
+  void collectRegInfo(HWModuleOp op, SmallVector<RegInfoCollect> &rInfo);
+
+  /// The function to override all modules in InstanceGraph based on a
+  /// depth-first algorithm.
+  void dfsOverrideModules(
+      MLIRContext *ctx,
```

Comment:
Isn't context here redundant with lots of other things pass in?

Reply:
There does seem to be a bit of redundancy here that I‘m not thinking about, I 'll remove it.